decay.go

```diff
@@ -0,0 +1,268 @@
+package connmgr
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/connmgr"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	"github.com/benbjohnson/clock"
+)
+
+// DefaultResolution is the default resolution of the decay tracker.
+var DefaultResolution = 1 * time.Minute
+
+// bumpCmd represents a bump command.
+type bumpCmd struct {
+	peer  peer.ID
+	tag   *decayingTag
+	delta int
+}
+
+// decayer tracks and manages all decaying tags and their values.
+type decayer struct {
+	cfg   *DecayerCfg
+	mgr   *BasicConnMgr
+	clock clock.Clock // for testing.
+
+	tagsMu    sync.Mutex
+	knownTags map[string]*decayingTag
+
+	// currRound is the current round we're in; guarded by atomic.
+	currRound uint64
+
+	// bumpCh queues bump commands to be processed by the loop.
+	bumpCh chan bumpCmd
+
+	// closure thingies.
+	closeCh chan struct{}
+	doneCh  chan struct{}
+	err     error
+}
+
+var _ connmgr.Decayer = (*decayer)(nil)
+
+// DecayerCfg is the configuration object for the Decayer.
+type DecayerCfg struct {
+	Resolution time.Duration
+	Clock      clock.Clock
+}
+
+// WithDefaults writes the default values on this DecayerConfig instance,
+// and returns itself for chainability.
+//
+//  cfg := (&DecayerCfg{}).WithDefaults()
+//  cfg.Resolution = 30 * time.Second
+//  t := NewDecayer(cfg, cm)
+func (cfg *DecayerCfg) WithDefaults() *DecayerCfg {
+	cfg.Resolution = DefaultResolution
+	return cfg
+}
+
+// NewDecayer creates a new decaying tag registry.
+func NewDecayer(cfg *DecayerCfg, mgr *BasicConnMgr) (*decayer, error) {
+	// use real time if the Clock in the config is nil.
+	if cfg.Clock == nil {
+		cfg.Clock = clock.New()
+	}
+
+	d := &decayer{
+		cfg:       cfg,
+		mgr:       mgr,
+		clock:     cfg.Clock,
+		knownTags: make(map[string]*decayingTag),
+		bumpCh:    make(chan bumpCmd, 128),
+		closeCh:   make(chan struct{}),
+		doneCh:    make(chan struct{}),
+	}
+
+	// kick things off.
+	go d.process()
+
+	return d, nil
+}
+
+// decayingTag represents a decaying tag, with an associated decay interval, a
+// decay function, and a bump function.
+type decayingTag struct {
+	trkr      *decayer
+	name      string
+	interval  time.Duration
+	nextRound uint64
+	decayFn   connmgr.DecayFn
+	bumpFn    connmgr.BumpFn
+}
+
+func (d *decayer) RegisterDecayingTag(name string, interval time.Duration, decayFn connmgr.DecayFn, bumpFn connmgr.BumpFn) (connmgr.DecayingTag, error) {
+	d.tagsMu.Lock()
+	defer d.tagsMu.Unlock()
+
+	tag, ok := d.knownTags[name]
+	if ok {
+		return nil, fmt.Errorf("decaying tag with name %s already exists", name)
+	}
+
+	if interval < d.cfg.Resolution {
+		err := fmt.Errorf("decay interval for %s (%s) is lower than tracker's resolution (%s)",
+			name,
+			interval,
+			d.cfg.Resolution)
+		return nil, err
+	}
+
+	if interval%d.cfg.Resolution != 0 {
```

Comment:
@raulk this seems problematic, since it's likely that the component registering the tag is not the same component that configured the `Decayer` resolution, and there's not an easy way to get the resolution out of the `Decayer` after it's been constructed. 

So if the `Decayer` was created with a non-default resolution, my component that just wants to register a tag will fail, and it can't easily discover a working value to use instead without parsing the error message.

That said, silently ignoring the loss of precision seems wrong, so I'm not sure what the right behavior is...

Reply:
Yeah, that's a good point. Let me think what we can do here.

Reply:
As of https://github.com/libp2p/go-libp2p-connmgr/pull/61/commits/b02e214d5b16c30c353c99f5ba8bd17b11f53293, we are more lenient:

* We now allow the user to request tags at an interval lower than the decayer's resolution, and we override the effective interval
to be the decayer's resolution. Users can now obtain the effective interval through the `Interval()` method on the decaying tag.

* We also allow tags to have any interval, not necessarily a multiple of the resolution any longer.